x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java

```diff
@@ -769,14 +769,16 @@ public void testGetAliasesAliasesOnlyPermissionIgnoreUnavailable() {
             .get();
         assertEquals(0, getAliasesResponse.getAliases().size());
 
-        // no manage_aliases privilege on non_authorized alias
-        getAliasesResponse = client.admin()
-            .indices()
-            .prepareGetAliases("non_authorized")
-            .addIndices("test_1")
-            .setIndicesOptions(IndicesOptions.lenientExpandOpen())
-            .get();
-        assertEquals(0, getAliasesResponse.getAliases().size());
+        // no manage_aliases privilege on `non_authorized` alias and ignore_unavailable doesn't influence aliases
+        assertThrowsAuthorizationException(
+            client.admin()
+                .indices()
+                .prepareGetAliases("non_authorized")
+                .addIndices("test_1")
+                .setIndicesOptions(IndicesOptions.lenientExpandOpen())::get,
+            GetAliasesAction.NAME,
+            "aliases_only"
+        );
```

Comment:
The `ignore_unavailable` request option does not currently work for "unavailable" aliases, see: https://github.com/elastic/elasticsearch/blob/844bb4cee573b4eb2b66fd2253d364ae7cf935fa/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L375 .

It worked in this particular test by coincidence.
In this particular test the `test_1` index is authorized but non-existent, so the request was previously authorized after the request was rewritten to empty. But now this is not the case anymore.
`test_1` is authorized, so it is not considered "unavailable" from Security's POV,

Reply:
From end-user's perspective, strictly speaking, this is a slight behaviour change:
* Status code from 404 to 403
* Response content from containing found aliases (and missing aliases) to unauthorized error response

Just want to call it out explicitly. But I don't think it is a real concern because the existing response is not a 200 anyway and it feels like an edge case.

Reply:
Yes, it's indeed a behavior change for an edge case.

Rereading this now, after a week, I realise I haven't done a good job explaining what's going on here.
Let me try again:

The get aliases action in this test uses the "lenient expand open" request option, which means that missing explicit indices are ignored (and that wildcards can expand to empty).
The `test_1` index does not exist, and it is the only index in the request, so when retrieving the `non_authorized` alias for it, because of the request option, this previously returned a 200 empty result, because there were ultimately no indices to return the aliases for.
But, after the change, Security does not care if `test_1` exists or not (Core does, and if Security lets the request through, given the request option, it will return the empty result). For Security, the `test_1` is authorized, but the alias `non_authorized` is not, hence it will then return a 403 because of the alias. For **indices**, the request option `ignore_unavailable` silently discards unauthorized index names, but the option never discards unauthorized **aliases** (but maybe it should?).

Therefore, clients will notice behavior differences for this API, when **all** the indices are authorized but missing, the request option tells to ignore missing indices, and there are unauthorized aliases (missing or not). Before they got a 200 empty, after they get a 403.
They got

Reply:
>  Before they got a 200 empty, after they get a 403.

Does it return 200 today? At REST layer, the response code is set to 404 if there is any explicitly requested alias not found.

https://github.com/elastic/elasticsearch/blob/c0e702c8365a068a8e0ff7290264e622158dea03/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java#L130-L133